tests/plugins/Qukeys/basic/test/testcase.cpp

```diff
@@ -32,182 +32,248 @@ constexpr KeyAddr key_addr_X{3, 2};
+    return expectReportWith(key, message);
+  }
+  uint32_t expectReportAfterCyclesWithout(size_t n, Key key,
+                                          std::string message = "") {
+    sim_.RunCycles(n);
+    return expectReportWithout(key, message);
+  }
+  uint32_t expectReportAfterMillis(uint32_t t,
+                                   std::set<Key> added_keys,
+                                   std::set<Key> removed_keys,
+                                   std::string message = "") {
+    sim_.RunForMillis(t);
+    return expectReport(added_keys, removed_keys, message);
+  }
+  uint32_t expectReportAfterMillis(uint32_t t, std::string message = "") {
+    return expectReportAfterMillis(t, std::set<Key>{}, std::set<Key>{},
+                                   message);
+  }
+  uint32_t expectReportAfterMillisWith(uint32_t t, Key key,
+                                       std::string message = "") {
+    sim_.RunForMillis(t);
+    return expectReportWith(key, message);
+  }
+  uint32_t expectReportAfterMillisWithout(uint32_t t, Key key,
+                                          std::string message = "") {
+    sim_.RunForMillis(t);
+    return expectReportWithout(key, message);
+  }
 };
 
 TEST_F(QukeysBasic, TapQukeyAlone) {
 
-  // Press `A`
-  sim_.Press(key_addr_A);
-
-  state_ = RunCycle();
-
-  ASSERT_EQ(state_->HIDReports()->Keyboard().size(), 0)
-      << "There should be no HID report after the qukey is pressed";
-  sim_.RunForMillis(20);
-  // Release `A`
-  sim_.Release(key_addr_A);
-
-  sim_.RunCycles(2);
-  state_ = RunCycle();
-
-  ASSERT_EQ(state_->HIDReports()->Keyboard().size(), 2)
-      << "There should be two HID reports after the release of a tapped qukey";
-
-  expected_keycodes_.insert(Key_A.getKeyCode());
-  EXPECT_THAT(state_->HIDReports()->Keyboard(0).ActiveKeycodes(),
-              ::testing::ElementsAreArray(expected_keycodes_))
-      << "The first report should include only `A`";
-
-  expected_keycodes_.erase(Key_A.getKeyCode());
-  EXPECT_THAT(state_->HIDReports()->Keyboard(1).ActiveKeycodes(),
-              ::testing::ElementsAreArray(expected_keycodes_))
-      << "The second report should be empty";
-
-  state_ = RunCycle();
-  ASSERT_EQ(state_->HIDReports()->Keyboard().size(), 0);
+  pressKey(10, key_addr_A);
+  releaseKey(50, key_addr_A);
+  expectReportAfterCyclesWith(
+      1, Key_A, "Report should contain only `A`");
+  expectReportAfterCyclesWithout(
+      1, Key_A, "Report should be empty");
+
+  sim_.RunForMillis(10);
+  auto state = RunCycle();
+
+  constexpr int expected_report_count = 2;
+  ASSERT_EQ(state->HIDReports()->Keyboard().size(), expected_report_count)
+      << "There should be two HID reports after a qukey is tapped";
+
+  for (int i = 0; i < expected_report_count; ++i) {
+    EXPECT_THAT(state->HIDReports()->Keyboard(i).ActiveKeycodes(),
+                ::testing::ElementsAreArray(expected_reports_[i].Keycodes()))
+        << expected_reports_[i].Message();
+  }
```

Comment:
When I was writing a series of `EXPECT_THAT()` statements, I did a lot of copying and pasting, and several times failed to update an index number. This is the reason for the inclusion of the failure message in the `expected_reports_[]` entries. I'm not sure how well this will work until I write a bunch of new testcases for Qukeys and OneShot, but I think that this will be a useful model for doing them clearly and with very few errors.